### PR TITLE
Show reachable scores in course statistics and fix tooltips of scores

### PR DESCRIPTION
**Pull request by sjagla27**
# Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Client: I added multiple screenshots/screencasts of my UI changes
- [X] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
Course statistics show (relative) scores even though there are still pending results.

### Description
Statistics now show the absolute points, reachable points (where pending results are not included), and total points. Reachable and total include their respective relative scores. The total diagram on the left now also doesn't include exercises that haven't been assessed yet.

### Steps for Testing

1. Log in to Artemis
2. Navigate to a Course Overview (prefereably one where the user has results for exercises)
3. Go to Statistics
4. Check whether Your points, Reachable points and Total points display the correct numbers.
-> in reachable included should be every task where there is a finished result, so assessed submissions and exercises not participated in.
4.1 Check if points are correctly displayed on the tooltip of the bar chart.
5. Add new exercises with fitting due dates and assessment due dates.
6. Participate in those exercises and assess them (or not).
7. Repeat step 4.

Tooltips of programming exercises were the main source of inconsistencies. Take an extra look if the tooltips are correct for those.

### Screenshots

Before:
Screenshot_2
After:
Screenshot_1

### CODE REVIEW

**Comment by Alexanderkrp** [comment]
Hello Sebastian

**Comment by Alexanderkrp** [comment]
Hello Sebastian
I just reviewed your PR and some things caught my attention:

- So I created two FileUpload Exercises, just to make sure everything works as expected, but as you can see the statistics show 4 reachable points for the test2 Exercise even tho the Due Date is over and I have not submitted anything. Thus, I should have 0 reachable points for this one. On the other hand I would expect both exercises to be displayed because you already know the total amount of points of both [8 Points in total] and there should be 4 reachable points for the Test Exercise with ID 2201 instead, since you are still able to work on this and reach 4 points.
image

image
image



- pretty much the same logic applies in the next example:
image
image
in my understanding it should be 20 reachable points since Text Exercise with ID 2114 has not been assessed yet AND you can still work on Text Exercise with ID 2150 both of which give 10 points.

- You could maybe adjust this and add another statistics which is called : Points to be assessed so that the student can see for which Exercises he has already made a submission and for which he still has to make one

I hope i can help you with my feedback. Apart from this it looks good 👍

**Comment by sjagla27** [comment]
@Alexanderkrp 
I think there is so kind of confusion what reachable means maybe that has to be clearer somehow. Reachable refers to all exercises that have an assessed result or the student did not participated in. 

First point: The exercises first show up in the statistics once their due date is over. There was nothing changed here. . Exercises the student did not participate should be included in reachable points because the student technically could have participated and gained those points.

Second point: Having exercises without due date screws with the statistics since they are on the one hand technically finished because they have no due date and thus should be shown in the statistics and on the other hand can still be worked on and have points to gain.

Third points: Points to be assessed are kind of the difference between total and reachable points so I dont think they need to be displayed separately.

**Comment by fde312** [comment]
image
File 1: participated in exercise and saved submission (not assessed yet)

File 2: did not participate at all (did not click on 'Start Exercise')
File 3: clicked on 'Start Exercise' and nothing else. It shows that I handed in a late participation


I have the same issue like @Alexanderkr when not participating (file 2) in an exercise. I would understand it that way that it shows me right now 0 as reachable score. 
Other than that it looks good to me :)

**Comment by krusche** [comment]
I'll close this for now to improve the overview on the PR page. Feel free to open it at a later point when you work on it again

File: src/main/webapp/app/overview/course-score-calculation.service.ts [hunk 1]
```diff
@@ -36,8 +39,15 @@ export class CourseScoreCalculationService {
                             score = 0;
                         }
                         absoluteScore = absoluteScore + score * this.SCORE_NORMALIZATION_VALUE * exercise.maxScore;
+                        reachableScore += exercise.maxScore;
                     }
                     presentationScore += participation.presentationScore !== undefined ? participation.presentationScore : 0;
+
+                    if (participation.initializationState === InitializationState.INITIALIZED && exercise.type !== ExerciseType.PROGRAMMING) {
+                        reachableScore += exercise.maxScore;
```
Comment (sleiss) [inline src/main/webapp/app/overview/course-score-calculation.service.ts:47]: How do you handle programming exercises? I cant find a specific condition that match programming exercises?

It seems to work, I just could not figure out why, maybe you could add a comment for that?

File: src/main/webapp/app/overview/course-statistics/course-statistics.component.ts [hunk 2]
```diff
@@ -343,28 +344,82 @@ export class CourseStatisticsComponent implements OnInit, OnDestroy {
...
+                groupedExercise.missedScores.tooltips.push(`${result.resultString} (${100 - result.score}%)`);
+            }
+
+            return;
         }
-        if (result.submission && result.submission.submissionExerciseType === SubmissionExerciseType.PROGRAMMING && (result.submission as ProgrammingSubmission).buildFailed) {
-            return null;
+
+        if (replaced.includes('points')) {
+            if (split.length === 2) {
```
Comment (sleiss) [inline src/main/webapp/app/overview/course-statistics/course-statistics.component.ts:379]: Could you please document what split.length == 2/4 corresponds to (which exercisetype)?

File: src/main/webapp/app/overview/course-statistics/course-statistics.component.ts [hunk 3]
```diff
@@ -343,28 +344,82 @@ export class CourseStatisticsComponent implements OnInit, OnDestroy {
...
+                );
+                return;
+            }
         }
-        if (result.resultString.indexOf('of') === -1) {
-            if (result.resultString.indexOf('points') === -1) {
-                return 0;
+
+        if (replaced.includes('passed')) {
+            if (split.length === 2) {
```
Comment (sleiss) [inline src/main/webapp/app/overview/course-statistics/course-statistics.component.ts:412]: Same as above.

File: src/main/webapp/app/overview/course-statistics/course-statistics.component.ts (hunk 2 above)
Comment (sjagla27) [reply src/main/webapp/app/overview/course-statistics/course-statistics.component.ts:381]: Done.

File: src/main/webapp/app/overview/course-statistics/course-statistics.component.ts (hunk 3 above)
Comment (sjagla27) [reply src/main/webapp/app/overview/course-statistics/course-statistics.component.ts:415]: Done.

File: src/main/webapp/app/overview/course-score-calculation.service.ts (hunk 1 above)
Comment (sjagla27) [reply src/main/webapp/app/overview/course-score-calculation.service.ts:48]: I believe programming exercises keep their initialized state even after they are done. Therefore the reachable score was added twice.

**Comment by sjagla27** [comment]
@sleiss 
Yeah I noticed the positioning as well. Will fix it in a different PR.

File: src/main/webapp/app/overview/course-statistics/course-statistics.component.ts [hunk 4]
```diff
@@ -424,6 +482,38 @@ export class CourseStatisticsComponent implements OnInit, OnDestroy {
         this.totalRelativeScore = this.calculateTotalScoreForTheCourse(RELATIVE_SCORE);
     }
 
+    calculateReachableScores() {
+        const quizzesTotalCurrentMaxScore = this.calculateScoreTypeForExerciseType(ExerciseType.QUIZ, REACHABLE_SCORE);
```
Comment (stefanwaldhauser) [inline src/main/webapp/app/overview/course-statistics/course-statistics.component.ts:486]: Why do you need all those constants? As far as I can see you you just assign them below to reachable scores?

File: src/main/webapp/app/overview/course-statistics/course-statistics.component.ts [hunk 5]
```diff
@@ -424,6 +482,38 @@ export class CourseStatisticsComponent implements OnInit, OnDestroy {
...
+        reachableScores[ExerciseType.PROGRAMMING] = programmingExerciseTotalCurrentMaxScore;
+        reachableScores[ExerciseType.MODELING] = modelingExerciseTotalCurrentMaxScore;
+        reachableScores[ExerciseType.TEXT] = textExerciseTotalCurrentMaxScore;
+        reachableScores[ExerciseType.FILE_UPLOAD] = fileUploadExerciseTotalCurrentMaxScore;
+        this.reachableScores = reachableScores;
+        this.reachableScore = this.calculateTotalScoreForTheCourse(REACHABLE_SCORE);
+    }
+
+    calculateCurrentRelativeScores(): void {
+        const quizzesCurrentRelativeScore = this.calculateScoreTypeForExerciseType(ExerciseType.QUIZ, CURRENT_RELATIVE_SCORE);
```
Comment (stefanwaldhauser) [inline src/main/webapp/app/overview/course-statistics/course-statistics.component.ts:502]: Why do you need all those constants? As far as I can see you you just assign them below?

File: src/main/webapp/app/overview/course-score-calculation.service.ts [hunk 6]
```diff
@@ -1,16 +1,18 @@
...
 import { Moment } from 'moment';
 import { StudentParticipation } from 'app/entities/participation/student-participation.model';
-import { Participation } from 'app/entities/participation/participation.model';
+import { InitializationState, Participation } from 'app/entities/participation/participation.model';
 
 export const ABSOLUTE_SCORE = 'absoluteScore';
 export const RELATIVE_SCORE = 'relativeScore';
 export const MAX_SCORE = 'maxScore';
 export const PRESENTATION_SCORE = 'presentationScore';
+export const REACHABLE_SCORE = 'reachableScore';
```
Comment (stefanwaldhauser) [inline src/main/webapp/app/overview/course-score-calculation.service.ts:14]: I guess this is not in the scope of this ticket but I am unsure why we use a Map<string,number> here and export the keys in this complicated fashion.  Would something like 
`interface CourseScores {
    absoluteScore: number;
    relativeScore: number;
}
`
not be much cleaner and more in the spirit of typescript?

**Comment by sleiss** [comment]
@kloessst 
> All five programming exercises have max points 5. I expected that the total points would be 25. One of those exercise was added with manual review initially. I changed the grading to automatic for this exercise to check if it fixes the problem but nothing changed.

I've just checked the course where you took the screenshot in and it seems that the 'manual review' exercise has a due date in 3 hours (from now). I've changed the due date to be in the past and then it is shown with 25 points for programming exercises (it still is in the past so you can simply check it here: https://artemistest.ase.in.tum.de/#/courses/293/statistics).

Can you confirm that this is the same for you?

**Pull request merged**